include/ofi_util.h

```diff
@@ -327,13 +327,28 @@ static inline void ofi_ep_rem_wr_cntr_inc(struct util_ep *ep)
 
 typedef void (*ofi_ep_cntr_inc_func)(struct util_ep *);
 
-static const ofi_ep_cntr_inc_func ofi_ep_cntr_inc_funcs[] = {
-	[FI_TRANSMIT] = ofi_ep_tx_cntr_inc,
-	[FI_RECV] = ofi_ep_rx_cntr_inc,
-	[FI_READ] = ofi_ep_rd_cntr_inc,
-	[FI_WRITE] = ofi_ep_wr_cntr_inc,
-	[FI_REMOTE_READ] = ofi_ep_rem_rd_cntr_inc,
-	[FI_REMOTE_WRITE] = ofi_ep_rem_wr_cntr_inc,
+static const ofi_ep_cntr_inc_func ofi_ep_tx_cntr_inc_funcs[] = {
+	[ofi_op_msg] = ofi_ep_tx_cntr_inc,
+	[ofi_op_tagged] = ofi_ep_tx_cntr_inc,
+	[ofi_op_read_req] = ofi_ep_rd_cntr_inc,
+	[ofi_op_read_rsp] = ofi_ep_rd_cntr_inc,
+	[ofi_op_write] = ofi_ep_wr_cntr_inc,
+	[ofi_op_write_rsp] = ofi_ep_wr_cntr_inc,
+	[ofi_op_atomic] = ofi_ep_wr_cntr_inc,
+	[ofi_op_atomic_fetch] = ofi_ep_rd_cntr_inc,
+	[ofi_op_atomic_compare] = ofi_ep_rd_cntr_inc,
+};
```

Comment:
Can you describe the intent of how these functions are used?  Specifically, are they called on the completion of a transmit operation with the requested opcode?

I'm confused because of the read/write response increments.  My brain wants those to either be no-ops or updates to remote counters.

Reply:
I refrained from making it a no-op mainly because of the shm provider's usage of these ops using the 2 different protocols. It uses ofi_op_write/ofi_op_read_req when it's sending an RMA that the peer needs to process (ie no CMA) and uses ofi_op_write_rsp/ofi_op_read/rsp when it's sending a notification to the peer that it did a write/read (with CMA).
With that aside, though, the general usage is as follows:
ofi_ep_tx_cntr_funcs : used to increment counters corresponding to ops initiated locally (all transmits)
ofi_ep_rx_cntr_funcs : used to increment counters corresponding to ops initiated remotely (all receives and remote RMAs)

Reply:
I'm still confused then.  Why doesn't rx write_rsp (received a write response?) increment the initiator write counter?

What my brains wants to see is:

Local side transmits op code X.  Transmit for op code X completes.  If X should update a counter, update counter N.  In the case X = response, I was expecting N to be a remote counter.

Similar on the receive side, if receiving X = response, I was expecting an update to the initiator counter.

Reply:
I'm not sure I follow your issue.

If a peer receives an RMA response, it will call
ofi_ep_rx_cntr_inc_func[ofI_write_rsp] -> ofi_rem_wri_cntr_inc
or
ofi_ep_rx_cntr_inc_funcs[ofi_read_rsp] -> ofi_rem_rd_cntr_inc

And the initiator of that op will call
ofi_ep_tx_cntr_inc_funcs[ofi_op_write_rsp] -> ofi_wr_cntr_inc
or
ofi_ep_tx_cntr_inc_funcs[ofi_op_read_rsp] -> ofi_rd_cntr_inc